core/src/main/java/org/elasticsearch/gateway/PrimaryShardAllocator.java

```diff
@@ -94,110 +97,155 @@ public PrimaryShardAllocator(Settings settings) {
+                logger.debug("[{}][{}]: allocating [{}] to [{}] on forced primary allocation",
+                             unassignedShard.index(), unassignedShard.id(), unassignedShard, nodeShardState.getNode());
+                return UnassignedShardDecision.yesDecision("",
+                    nodeShardState.getNode().getId(),
+                    nodeShardState.allocationId(),
+                    (explain ? buildNodeDecisions(nodesToForceAllocate) : null));
+            } else if (nodesToForceAllocate.throttleNodeShards.isEmpty() == false) {
+                logger.debug("[{}][{}]: throttling allocation [{}] to [{}] on forced primary allocation",
+                             unassignedShard.index(), unassignedShard.id(), unassignedShard, nodesToForceAllocate.throttleNodeShards);
+                return UnassignedShardDecision.throttleDecision(
+                    "every node to which the shard can be force allocated is busy with other recoveries, so we are throttling allocation",
+                    (explain ? buildNodeDecisions(nodesToForceAllocate) : null));
             } else {
-                // we are throttling this, but we have enough to allocate to this node, ignore it for now
-                logger.debug("[{}][{}]: throttling allocation [{}] to [{}] on primary allocation", shard.index(), shard.id(), shard, nodesToAllocate.throttleNodeShards);
-                unassignedIterator.removeAndIgnore(AllocationStatus.DECIDERS_THROTTLED, allocation.changes());
+                logger.debug("[{}][{}]: forced primary allocation denied [{}]",
+                             unassignedShard.index(), unassignedShard.id(), unassignedShard);
+                return UnassignedShardDecision.noDecision(AllocationStatus.DECIDERS_NO,
+                    "all nodes that hold a valid shard copy received a NO decision, and force allocation was not permitted either",
+                    (explain ? buildNodeDecisions(nodesToForceAllocate) : null));
             }
+        } else {
+            // we are throttling this, since we are allowed to allocate to this node but there are enough allocations
+            // taking place on the node currently, ignore it for now
+            logger.debug("[{}][{}]: throttling allocation [{}] to [{}] on primary allocation",
+                         unassignedShard.index(), unassignedShard.id(), unassignedShard, nodesToAllocate.throttleNodeShards);
+            return UnassignedShardDecision.throttleDecision(
+                "every node to which the shard can be allocated is busy with other recoveries, so we are throttling allocation",
+                (explain ? buildNodeDecisions(nodesToAllocate) : null));
+        }
+    }
+
+    /**
+     * Builds a map of nodes to the corresponding allocation decisions for those nodes.
+     */
+    private Map<String, Decision> buildNodeDecisions(NodesToAllocate nodesToAllocate) {
```

Comment:
I think rather than the inline `(explain ? buildNodeDecisions(...) : null)` check that is repeated many times, it'd be better to pass the `explain` parameter to this function (which I think can be static?) and always call it, and then this method can return `null` if `explain` is false